cpp/velox/shuffle/VeloxShuffleWriter.cc

```diff
@@ -1134,50 +1152,55 @@ arrow::Status VeloxShuffleWriter::splitFixedWidthValueBuffer(const velox::RowVec
     return arrow::Status::OK();
   }
 
+  inline bool VeloxShuffleWriter::needRealloc(uint32_t partitionId, uint64_t newSize) {
+    auto currentBufferSize = partition2BufferSize_[partitionId];
+    return newSize > (1 + options_.buffer_realloc_threshold) * currentBufferSize ||
+        newSize < (1 - options_.buffer_realloc_threshold) * currentBufferSize;
+  }
+
   uint32_t VeloxShuffleWriter::calculatePartitionBufferSize(const velox::RowVector& rv) {
-    uint32_t sizePerRow = 0;
+    uint32_t bytesPerRow = 0;
     auto numRows = rv.size();
-    for (size_t i = fixedWidthColumnCount_; i < simpleColumnIndices_.size(); ++i) {
-      auto index = i - fixedWidthColumnCount_;
-      if (binaryArrayEmpiricalSize_[index] == 0) {
-        auto column = rv.childAt(simpleColumnIndices_[i]);
-        auto stringViewColumn = column->asFlatVector<velox::StringView>();
-        assert(stringViewColumn);
-
-        // accumulate length
-        uint64_t length = stringViewColumn->values()->size();
-        for (auto& buffer : stringViewColumn->stringBuffers()) {
-          length += buffer->size();
-        }
-
-        binaryArrayEmpiricalSize_[index] = length % numRows == 0 ? length / numRows : length / numRows + 1;
+    // Calculate average size bytes (bytes per row) for each binary array.
+    std::vector<uint64_t> binaryArrayAvgBytesPerRow(binaryColumnIndices_.size());
+    for (size_t i = 0; i < binaryColumnIndices_.size(); ++i) {
+      auto column = rv.childAt(binaryColumnIndices_[i]);
+      auto stringViewColumn = column->asFlatVector<velox::StringView>();
+      assert(stringViewColumn);
+
+      uint64_t binarySizeBytes = stringViewColumn->values()->size();
+      for (auto& buffer : stringViewColumn->stringBuffers()) {
+        binarySizeBytes += buffer->size();
       }
-    }
 
-    VS_PRINT_VECTOR_MAPPING(binaryArrayEmpiricalSize_);
+      binaryArrayTotalSizeBytes_[i] += binarySizeBytes;
+      binaryArrayAvgBytesPerRow[i] = binaryArrayTotalSizeBytes_[i] / (totalInputNumRows_ + numRows);
+      bytesPerRow += binaryArrayAvgBytesPerRow[i];
+    }
 
-    sizePerRow = std::accumulate(binaryArrayEmpiricalSize_.begin(), binaryArrayEmpiricalSize_.end(), 0);
+    VS_PRINT_VECTOR_MAPPING(binaryArrayAvgBytesPerRow);
 
     for (size_t col = 0; col < simpleColumnIndices_.size(); ++col) {
       auto colIdx = simpleColumnIndices_[col];
       // `bool(1) >> 3` gets 0, so +7
-      sizePerRow += ((arrow::bit_width(arrowColumnTypes_[colIdx]->id()) + 7) >> 3);
+      bytesPerRow += ((arrow::bit_width(arrowColumnTypes_[colIdx]->id()) + 7) >> 3);
     }
 
-    VS_PRINTLF(sizePerRow);
+    VS_PRINTLF(bytesPerRow);
```

Comment:
Let's use bytesPerRow = estimateFlatSize()/rv.size(); to calculate directly. It's a little over estimate the real size but it's fine. The function doesn't itereate the string_view array in binary column, instead it use buffer_array which several or all strings may share the same buffer.